prov/sharp scaffolding

grom72 opened the pull request:
Prov sharp scaffolding

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/libfabric/8)
<!-- Reviewable:end -->

grom72 committed:
prov/sharp: SHARP provider scaffolding

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

grom72 committed:
prov/rxm: collective offload capabilities reported if offload provider is available

* FOR TEST PURPOSE ONLY - TEMPORARY SOLUTION *
To let all collective tests pass let's force rxm to provide only offload collective
domain capabilities. THIS IS ONLY FOR TEST PPURPOSE. In the future, tests for
offload provider shall be executed with OFI_OFFLOAD_PROV_ONLY flag set in
fi_query_collective() call.

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

grom72 committed:
prov/sharp: Dummy implementation of fi_barrier() and fi_allreduce()

Both collective operation implemented ba colling peer collective operation and
transparently passing completion back to peer CQ

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

ldorau reviewed (changes requested):
> Reviewed 4 of 24 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/pmem/libfabric/8)* status: 4 of 24 files reviewed, 2 unresolved discussions (waiting on @grom72)

___
*[`Makefile.am` line 6 at r1](https://reviewable.io/reviews/pmem/libfabric/8#-NJEoLn_C_fEFYrOqp1K:-NJEoLn_C_fEFYrOqp1L:b-c6hy4h) ([raw file](https://github.com/pmem/libfabric/blob/fa83a1dd95fe3893b4fc0aaaf4eaac0d491c0a0e/Makefile.am#L6)):*
> ```Makefile
> # Copyright (c) 2018 Amazon.com, Inc. or its affiliates. All rights reserved.
> # (C) Copyright 2020 Hewlett Packard Enterprise Development LP
> # Copyright (c) 2022 Intel Corporation. All right reserved.
> ```

There is already Intel Copyright 3 lines above.



_Code quote:_
```Makefile
# Copyright (c) 2022 Intel Corporation. All right reserved.
```
___
*[`README.md` line 219 at r1](https://reviewable.io/reviews/pmem/libfabric/8#-NJEof-GC762ZJfyXbTN:-NJEof-GC762ZJfyXbTO:blrryw3) ([raw file](https://github.com/pmem/libfabric/blob/fa83a1dd95fe3893b4fc0aaaf4eaac0d491c0a0e/README.md#L219)):*
> ```Markdown
> 
> The `off_sharp` provider is an utility provider that supports collective endpoints utilizing
> SHARP protocol for barier and allreduce operations.
> ```

... for the barier and the allreduce ... 



_Code quote:_
```Markdown
for barier and allreduce
```


<!-- Sent from Reviewable.io -->

ldorau reviewed (changes requested):
> Reviewed 2 of 24 files at r1.
> *[Reviewable](https://reviewable.io/reviews/pmem/libfabric/8)* status: 6 of 24 files reviewed, 3 unresolved discussions (waiting on @grom72)

___
*[`include/ofi_sharp.h` line 70 at r1](https://reviewable.io/reviews/pmem/libfabric/8#-NJF0Bdc2rBPOmJhMD9j:-NJF0Bdc2rBPOmJhMD9k:b-y22m5) ([raw file](https://github.com/pmem/libfabric/blob/fa83a1dd95fe3893b4fc0aaaf4eaac0d491c0a0e/include/ofi_sharp.h#L70)):*
> ```objc
> #endif
> 
> #endif /* _OFI_SHM_H_ */
> ```

`_OFI_SHARP_H_`



_Code quote:_
```objc
_OFI_SHM_H_
```


<!-- Sent from Reviewable.io -->

grom72 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/pmem/libfabric/8)* status: 4 of 24 files reviewed, 3 unresolved discussions (waiting on @ldorau)

___
*[`README.md` line 219 at r1](https://reviewable.io/reviews/pmem/libfabric/8#-NJEof-GC762ZJfyXbTN:-NJF3hEqDrg8yNG_tlGk:b-896fix) ([raw file](https://github.com/pmem/libfabric/blob/fa83a1dd95fe3893b4fc0aaaf4eaac0d491c0a0e/README.md#L219)):*
<details><summary><i>Previously, ldorau (Lukasz Dorau) wrote…</i></summary><blockquote>

... for the barier and the allreduce ... 
</blockquote></details>

Done.
___
*[`include/ofi_sharp.h` line 70 at r1](https://reviewable.io/reviews/pmem/libfabric/8#-NJF0Bdc2rBPOmJhMD9j:-NJF3fQ70cBwhUBUWf0u:b-896fix) ([raw file](https://github.com/pmem/libfabric/blob/fa83a1dd95fe3893b4fc0aaaf4eaac0d491c0a0e/include/ofi_sharp.h#L70)):*
<details><summary><i>Previously, ldorau (Lukasz Dorau) wrote…</i></summary><blockquote>

`_OFI_SHARP_H_`
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

grom72 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/pmem/libfabric/8)* status: 3 of 24 files reviewed, 3 unresolved discussions (waiting on @ldorau)

___
*[`Makefile.am` line 6 at r1](https://reviewable.io/reviews/pmem/libfabric/8#-NJEoLn_C_fEFYrOqp1K:-NJFB4bxAZEbLiq9OAdw:b-896fix) ([raw file](https://github.com/pmem/libfabric/blob/fa83a1dd95fe3893b4fc0aaaf4eaac0d491c0a0e/Makefile.am#L6)):*
<details><summary><i>Previously, ldorau (Lukasz Dorau) wrote…</i></summary><blockquote>

There is already Intel Copyright 3 lines above.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ldorau reviewed (changes requested):
> Reviewed 2 of 3 files at r2, 1 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/pmem/libfabric/8)* status: 6 of 24 files reviewed, 1 unresolved discussion (waiting on @grom72)

___
*[`Makefile.am` line 3 at r3](https://reviewable.io/reviews/pmem/libfabric/8#-NJFFdH_BuJ4Gfm3LvAh:-NJFFdH_BuJ4Gfm3LvAi:b-gimt4y) ([raw file](https://github.com/pmem/libfabric/blob/6f2c738b6f399945efb8f8ac6fc0ed51157cc1ea/Makefile.am#L3)):*
> ```Makefile
> #
> # Copyright (c) 2016 Cisco Systems, Inc.  All rights reserved.
> # Copyright (c) 2017-2018, 2022 Intel Corporation, Inc. All right reserved.
> ```

I think it should be `2017-2022`



_Code quote:_
```Makefile
2017-2018, 2022
```


<!-- Sent from Reviewable.io -->

ldorau reviewed (changes requested):
> Reviewed 4 of 24 files at r1.
> *[Reviewable](https://reviewable.io/reviews/pmem/libfabric/8)* status: 10 of 24 files reviewed, 8 unresolved discussions (waiting on @grom72)

___
*[`man/fi_sharp.7.md` line 15 at r3](https://reviewable.io/reviews/pmem/libfabric/8#-NJFGFknASCG_pTLB9aj:-NJFGFknASCG_pTLB9ak:b-o40y2m) ([raw file](https://github.com/pmem/libfabric/blob/6f2c738b6f399945efb8f8ac6fc0ed51157cc1ea/man/fi_sharp.7.md#L15)):*
> ```Markdown
> 
> The SHARP provider is a collectives offload provider that can be used on Linux
> systems supporting SHARP protocol.
> ```

the SHARP protocol



_Code quote:_
```Markdown
SHARP protocol
```
___
*[`man/fi_sharp.7.md` line 19 at r3](https://reviewable.io/reviews/pmem/libfabric/8#-NJFGYq_7AwN9ewEIreB:-NJFGYq_7AwN9ewEIreC:b-81ncwr) ([raw file](https://github.com/pmem/libfabric/blob/6f2c738b6f399945efb8f8ac6fc0ed51157cc1ea/man/fi_sharp.7.md#L19)):*
> ```Markdown
> # SUPPORTED FEATURES
> 
> This release contains an initial implementation of the SHM provider that
> ```

SHARP



_Code quote:_
```Markdown
SHM
```
___
*[`man/fi_sharp.7.md` line 23 at r3](https://reviewable.io/reviews/pmem/libfabric/8#-NJFGoVbEZMRcXesMtVm:-NJFGoVbEZMRcXesMtVn:ba5gb9) ([raw file](https://github.com/pmem/libfabric/blob/6f2c738b6f399945efb8f8ac6fc0ed51157cc1ea/man/fi_sharp.7.md#L23)):*
> ```Markdown
> 
> *Endpoint types*
> : The provider supports only endpoint type *FI_EP_COLLECTIVE*.
> ```

`only one endpoint type: *FI_EP_COLLECTIVE*.`
or
`only the *FI_EP_COLLECTIVE* endpoint type.`



_Code quote:_
```Markdown
only endpoint type *FI_EP_COLLECTIVE*.
```
___
*[`man/fi_sharp.7.md` line 26 at r3](https://reviewable.io/reviews/pmem/libfabric/8#-NJFH-R14Cmszs1oMgym:-NJFH-R14Cmszs1oMgyn:b2ni08b) ([raw file](https://github.com/pmem/libfabric/blob/6f2c738b6f399945efb8f8ac6fc0ed51157cc1ea/man/fi_sharp.7.md#L26)):*
> ```Markdown
> 
> *Endpoint capabilities*
> : Endpoints cna support only fi_barrier and fi_allreduce operations.
> ```

can



_Code quote:_
```Markdown
cna
```
___
*[`man/fi_sharp.7.md` line 41 at r3](https://reviewable.io/reviews/pmem/libfabric/8#-NJFHHnx4vaJWgoYmgT0:-NJFHHnx4vaJWgoYmgT1:b-63o8s5) ([raw file](https://github.com/pmem/libfabric/blob/6f2c738b6f399945efb8f8ac6fc0ed51157cc1ea/man/fi_sharp.7.md#L41)):*
> ```Markdown
> 
> *MR registration mode*
>   The provider implements FI_MR_VIRT_ADDR memory mode.
> ```

the FI_MR_VIRT_ADDR memory mode



_Code quote:_
```Markdown
FI_MR_VIRT_ADDR memory mode
```
___
*[`man/fi_sharp.7.md` line 50 at r3](https://reviewable.io/reviews/pmem/libfabric/8#-NJFHTAoBg2VdNzBB7WD:-NJFHTAoBg2VdNzBB7WE:b-i1oxqc) ([raw file](https://github.com/pmem/libfabric/blob/6f2c738b6f399945efb8f8ac6fc0ed51157cc1ea/man/fi_sharp.7.md#L50)):*
> ```Markdown
> The SHARP provider has hard-coded maximums for supported queue sizes and data
> transfers.  These values are reflected in the related fabric attribute
> structures
> ```

`structures.` - dot at the end



_Code quote:_
```Markdown
structures
```
___
*[`man/fi_sharp.7.md` line 56 at r3](https://reviewable.io/reviews/pmem/libfabric/8#-NJFHdJr7mp5fU0krIA_:-NJFHdJr7mp5fU0krIAa:b7r8po9) ([raw file](https://github.com/pmem/libfabric/blob/6f2c738b6f399945efb8f8ac6fc0ed51157cc1ea/man/fi_sharp.7.md#L56)):*
> ```Markdown
> # RUNTIME PARAMETERS
> 
> The *SHARP* provider checks for the following environment variables:
> ```

Please be consistent: `The *SHARP* provider` or `The SHARP provider` - the same in all places



_Code quote:_
```Markdown
The *SHARP* provider
```


<!-- Sent from Reviewable.io -->

ldorau reviewed (commented):
> Reviewed 1 of 24 files at r1.
> *[Reviewable](https://reviewable.io/reviews/pmem/libfabric/8)* status: 11 of 24 files reviewed, 8 unresolved discussions (waiting on @grom72)



<!-- Sent from Reviewable.io -->

grom72 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/pmem/libfabric/8)* status: 9 of 24 files reviewed, 8 unresolved discussions (waiting on @ldorau)

___
*[`man/fi_sharp.7.md` line 15 at r3](https://reviewable.io/reviews/pmem/libfabric/8#-NJFGFknASCG_pTLB9aj:-NJFYu0n-Aqfw3WuNTsF:b-896fix) ([raw file](https://github.com/pmem/libfabric/blob/6f2c738b6f399945efb8f8ac6fc0ed51157cc1ea/man/fi_sharp.7.md#L15)):*
<details><summary><i>Previously, ldorau (Lukasz Dorau) wrote…</i></summary><blockquote>

the SHARP protocol
</blockquote></details>

Done.



_Code quote:_
```Markdown
 SHARP protocol.
```
___
*[`man/fi_sharp.7.md` line 19 at r3](https://reviewable.io/reviews/pmem/libfabric/8#-NJFGYq_7AwN9ewEIreB:-NJFZGxdE1kuBLU-sUqA:b-896fix) ([raw file](https://github.com/pmem/libfabric/blob/6f2c738b6f399945efb8f8ac6fc0ed51157cc1ea/man/fi_sharp.7.md#L19)):*
<details><summary><i>Previously, ldorau (Lukasz Dorau) wrote…</i></summary><blockquote>

SHARP
</blockquote></details>

Done.
___
*[`man/fi_sharp.7.md` line 23 at r3](https://reviewable.io/reviews/pmem/libfabric/8#-NJFGoVbEZMRcXesMtVm:-NJFZGZV-u5B7O5gWL5N:b-896fix) ([raw file](https://github.com/pmem/libfabric/blob/6f2c738b6f399945efb8f8ac6fc0ed51157cc1ea/man/fi_sharp.7.md#L23)):*
<details><summary><i>Previously, ldorau (Lukasz Dorau) wrote…</i></summary><blockquote>

`only one endpoint type: *FI_EP_COLLECTIVE*.`
or
`only the *FI_EP_COLLECTIVE* endpoint type.`
</blockquote></details>

Done.
___
*[`man/fi_sharp.7.md` line 26 at r3](https://reviewable.io/reviews/pmem/libfabric/8#-NJFH-R14Cmszs1oMgym:-NJFba6x56XrpKu3Vm4E:b-896fix) ([raw file](https://github.com/pmem/libfabric/blob/6f2c738b6f399945efb8f8ac6fc0ed51157cc1ea/man/fi_sharp.7.md#L26)):*
<details><summary><i>Previously, ldorau (Lukasz Dorau) wrote…</i></summary><blockquote>

can
</blockquote></details>

Done.
___
*[`man/fi_sharp.7.md` line 41 at r3](https://reviewable.io/reviews/pmem/libfabric/8#-NJFHHnx4vaJWgoYmgT0:-NJFbboU1Oa0Sfgd7jHh:b-896fix) ([raw file](https://github.com/pmem/libfabric/blob/6f2c738b6f399945efb8f8ac6fc0ed51157cc1ea/man/fi_sharp.7.md#L41)):*
<details><summary><i>Previously, ldorau (Lukasz Dorau) wrote…</i></summary><blockquote>

the FI_MR_VIRT_ADDR memory mode
</blockquote></details>

Done.
___
*[`man/fi_sharp.7.md` line 50 at r3](https://reviewable.io/reviews/pmem/libfabric/8#-NJFHTAoBg2VdNzBB7WD:-NJF_aLx1_3QRVbhvGb_:b-896fix) ([raw file](https://github.com/pmem/libfabric/blob/6f2c738b6f399945efb8f8ac6fc0ed51157cc1ea/man/fi_sharp.7.md#L50)):*
<details><summary><i>Previously, ldorau (Lukasz Dorau) wrote…</i></summary><blockquote>

`structures.` - dot at the end
</blockquote></details>

Done.
___
*[`man/fi_sharp.7.md` line 56 at r3](https://reviewable.io/reviews/pmem/libfabric/8#-NJFHdJr7mp5fU0krIA_:-NJF__fd6hHYKFqsxzYf:b-896fix) ([raw file](https://github.com/pmem/libfabric/blob/6f2c738b6f399945efb8f8ac6fc0ed51157cc1ea/man/fi_sharp.7.md#L56)):*
<details><summary><i>Previously, ldorau (Lukasz Dorau) wrote…</i></summary><blockquote>

Please be consistent: `The *SHARP* provider` or `The SHARP provider` - the same in all places
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ldorau reviewed (commented):
> Reviewed 3 of 24 files at r1, 1 of 2 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/pmem/libfabric/8)* status: 15 of 24 files reviewed, 1 unresolved discussion (waiting on @grom72)



<!-- Sent from Reviewable.io -->

grom72 committed:
prov/sharp SHARP coll driver access API scaffolding

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

grom72 committed:
prov/sharp: Sharp provider TODO list

TODO.md files defines next development steps. The list is ordered by priority.

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

ldorau reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/pmem/libfabric/8)* status: 14 of 34 files reviewed, 2 unresolved discussions (waiting on @grom72)

___
*[`prov/sharp/Makefile.include` line 11 at r8](https://reviewable.io/reviews/pmem/libfabric/8#-NJLQtbY2UWYlQFt16yu:-NJLQtbY2UWYlQFt16yv:b4v1yxu) ([raw file](https://github.com/pmem/libfabric/blob/ac48ad7553144b9126aeeca6a794147a17198bb9/prov/sharp/Makefile.include#L11)):*
> ```
> 	prov/sharp/src/sharp_eq.c		\
> 	prov/sharp/src/sharp_ep.c		\
> 	prov/sharp/src/sharp_cq.c		\
> ```

Fix alignment, please (screenshot from vim):
![image.png](https://files.reviewableusercontent.io/images/reviewable/8492895/fH-Gd2C4vMXfylF2jPCQX91c/image.png)



_Code quote:_
```
_sharp_files = \
	include/ofi_sharp.h				\
	prov/sharp/src/sharp.h			\
	prov/sharp/src/sharp_attr.c		\
	prov/sharp/src/sharp_init.c		\
	prov/sharp/src/sharp_fabric.c	\
	prov/sharp/src/sharp_domain.c	\
	prov/sharp/src/sharp_eq.c		\
	prov/sharp/src/sharp_ep.c		\
	prov/sharp/src/sharp_cq.c		\
	prov/sharp/src/sharp_coll.c		\
	prov/sharp/src/sharp_progress.c
```


<!-- Sent from Reviewable.io -->

grom72 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/pmem/libfabric/8)* status: 12 of 34 files reviewed, 2 unresolved discussions (waiting on @ldorau)

___
*[`Makefile.am` line 3 at r3](https://reviewable.io/reviews/pmem/libfabric/8#-NJFFdH_BuJ4Gfm3LvAh:-NJPFcP61OG8kf-FRTnX:b-896fix) ([raw file](https://github.com/pmem/libfabric/blob/6f2c738b6f399945efb8f8ac6fc0ed51157cc1ea/Makefile.am#L3)):*
<details><summary><i>Previously, ldorau (Lukasz Dorau) wrote…</i></summary><blockquote>

I think it should be `2017-2022`
</blockquote></details>

Done.
___
*[`prov/sharp/Makefile.include` line 11 at r8](https://reviewable.io/reviews/pmem/libfabric/8#-NJLQtbY2UWYlQFt16yu:-NJPFjn1B6Uj43bWKqs6:b-896fix) ([raw file](https://github.com/pmem/libfabric/blob/ac48ad7553144b9126aeeca6a794147a17198bb9/prov/sharp/Makefile.include#L11)):*
<details><summary><i>Previously, ldorau (Lukasz Dorau) wrote…</i></summary><blockquote>

Fix alignment, please (screenshot from vim):
![image.png](https://files.reviewableusercontent.io/images/reviewable/8492895/fH-Gd2C4vMXfylF2jPCQX91c/image.png)

</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ldorau reviewed (commented):
> Reviewed 1 of 1 files at r8, 2 of 3 files at r9.
> *[Reviewable](https://reviewable.io/reviews/pmem/libfabric/8)* status: 14 of 34 files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

ldorau committed:
prov/sharp: Add mocks for sharp_coll_init and sharp_coll_finalize

Add mocks for sharp_coll_init and sharp_coll_finalize.

Signed-off-by: Lukasz Dorau <lukasz.dorau@intel.com>

ldorau committed:
prov/sharp: sharp provider can be mocked

In order to mock SHARP functions run:
$ ./configure --enable-sharp=mocked
or
$ ./configure --enable-sharp=auto

Running:
$ ./configure --enable-sharp=yes
make use of SHARP libraries instead of mocks.

Signed-off-by: Lukasz Dorau <lukasz.dorau@intel.com>

ldorau reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/pmem/libfabric/8)* status: 13 of 35 files reviewed, 8 unresolved discussions (waiting on @grom72)

___
*[`prov/sharp/src/sharp_cq.c` line 87 at r10](https://reviewable.io/reviews/pmem/libfabric/8#-NJoLc4w4_U4FwKvWQ4M:-NJoLc4w4_U4FwKvWQ4N:b-3vk3a0) ([raw file](https://github.com/pmem/libfabric/blob/f6743afdaa3c78b3303eeca1bca3b8062dc8cf33/prov/sharp/src/sharp_cq.c#L87)):*
> ```C
> 	if (!attr || !(attr->flags & FI_PEER)) {
> 		FI_WARN(provider, FI_LOG_CORE, "FI_PEER flag required\n");
>                 return -EINVAL;
> ```

`FI_EINVAL`



_Code quote:_
```C
EINVAL
```
___
*[`prov/sharp/src/sharp_cq.c` line 92 at r10](https://reviewable.io/reviews/pmem/libfabric/8#-NJoLeoD--5mZMOI5bqK:-NJoLeoD--5mZMOI5bqL:b3kqx) ([raw file](https://github.com/pmem/libfabric/blob/f6743afdaa3c78b3303eeca1bca3b8062dc8cf33/prov/sharp/src/sharp_cq.c#L92)):*
> ```C
> 	if (!peer_context || peer_context->size < sizeof(*peer_context)) {
> 		FI_WARN(provider, FI_LOG_CORE, "invalid peer CQ context\n");
>                 return -EINVAL;
> ```

.
___
*[`prov/sharp/src/sharp_domain.c` line 169 at r10](https://reviewable.io/reviews/pmem/libfabric/8#-NJoLgeUBilYwAGkA0sB:-NJoLgeV9-zaOPEDeKnn:b3kqx) ([raw file](https://github.com/pmem/libfabric/blob/f6743afdaa3c78b3303eeca1bca3b8062dc8cf33/prov/sharp/src/sharp_domain.c#L169)):*
> ```C
> 		FI_WARN(&sharp_prov, FI_LOG_CORE,
> 			"FI_PEER flag required\n");
> 		return -EINVAL;
> ```

.
___
*[`prov/sharp/src/sharp_domain.c` line 175 at r10](https://reviewable.io/reviews/pmem/libfabric/8#-NJoLhgV-AD5R-YdeJ5G:-NJoLhgV-AD5R-YdeJ5H:b3kqx) ([raw file](https://github.com/pmem/libfabric/blob/f6743afdaa3c78b3303eeca1bca3b8062dc8cf33/prov/sharp/src/sharp_domain.c#L175)):*
> ```C
> 		FI_WARN(&sharp_prov, FI_LOG_CORE,
> 			"Invalid peer domain context\n");
> 		return -EINVAL;
> ```

.
___
*[`prov/sharp/src/sharp_ep.c` line 188 at r10](https://reviewable.io/reviews/pmem/libfabric/8#-NJoLjMDCWEPQkuX6HA6:-NJoLjMDCWEPQkuX6HA7:b3kqx) ([raw file](https://github.com/pmem/libfabric/blob/f6743afdaa3c78b3303eeca1bca3b8062dc8cf33/prov/sharp/src/sharp_ep.c#L188)):*
> ```C
> 		FI_WARN(&sharp_prov, FI_LOG_CORE,
> 			"FI_PEER_TRANSFER mode required\n");
> 		return -EINVAL;
> ```

.
___
*[`prov/sharp/src/sharp_ep.c` line 194 at r10](https://reviewable.io/reviews/pmem/libfabric/8#-NJoLkSc9I5vADiZG-9G:-NJoLkSc9I5vADiZG-9H:b3kqx) ([raw file](https://github.com/pmem/libfabric/blob/f6743afdaa3c78b3303eeca1bca3b8062dc8cf33/prov/sharp/src/sharp_ep.c#L194)):*
> ```C
> 		FI_WARN(&sharp_prov, FI_LOG_CORE,
> 			"Invalid peer transfer context\n");
> 		return -EINVAL;
> ```

.
___
*[`prov/sharp/src/sharp_eq.c` line 76 at r10](https://reviewable.io/reviews/pmem/libfabric/8#-NJoLm778G_nuiYxA3Lx:-NJoLm778G_nuiYxA3Ly:b3kqx) ([raw file](https://github.com/pmem/libfabric/blob/f6743afdaa3c78b3303eeca1bca3b8062dc8cf33/prov/sharp/src/sharp_eq.c#L76)):*
> ```C
> 	if (!attr || !(attr->flags & FI_PEER)) {
> 		FI_WARN(&sharp_prov, FI_LOG_CORE, "FI_PEER flag required\n");
>                 return -EINVAL;
> ```

.
___
*[`prov/sharp/src/sharp_eq.c` line 81 at r10](https://reviewable.io/reviews/pmem/libfabric/8#-NJoLnDq5Yq27FnVo0UG:-NJoLnDq5Yq27FnVo0UH:b3kqx) ([raw file](https://github.com/pmem/libfabric/blob/f6743afdaa3c78b3303eeca1bca3b8062dc8cf33/prov/sharp/src/sharp_eq.c#L81)):*
> ```C
> 	if (!peer_context || peer_context->size < sizeof(*peer_context)) {
> 		FI_WARN(&sharp_prov, FI_LOG_CORE, "invalid peer EQ context\n");
>                 return -EINVAL;
> ```

.


<!-- Sent from Reviewable.io -->

grom72 committed:
prov/sharp: Dummy implementation of fi_join_collective()

fi_join_collective implementation with usage of coll_op's state machine

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

grom72 reviewed (commented):
> Reviewed 3 of 24 files at r1, 1 of 3 files at r2, 2 of 2 files at r4, 1 of 3 files at r9, 1 of 5 files at r12.
> *[Reviewable](https://reviewable.io/reviews/pmem/libfabric/8)* status: 12 of 37 files reviewed, 8 unresolved discussions (waiting on @ldorau)

___
*[`prov/sharp/src/sharp_cq.c` line 87 at r10](https://reviewable.io/reviews/pmem/libfabric/8#-NJoLc4w4_U4FwKvWQ4M:-NKSTp0A-5JqKqUImuzm:b-896fix) ([raw file](https://github.com/pmem/libfabric/blob/f6743afdaa3c78b3303eeca1bca3b8062dc8cf33/prov/sharp/src/sharp_cq.c#L87)):*
<details><summary><i>Previously, ldorau (Lukasz Dorau) wrote…</i></summary><blockquote>

`FI_EINVAL`
</blockquote></details>

Done.
___
*[`prov/sharp/src/sharp_cq.c` line 92 at r10](https://reviewable.io/reviews/pmem/libfabric/8#-NJoLeoD--5mZMOI5bqK:-NKSTq704ZWgHfed81TC:b-896fix) ([raw file](https://github.com/pmem/libfabric/blob/f6743afdaa3c78b3303eeca1bca3b8062dc8cf33/prov/sharp/src/sharp_cq.c#L92)):*
<details><summary><i>Previously, ldorau (Lukasz Dorau) wrote…</i></summary><blockquote>

.
</blockquote></details>

Done.
___
*[`prov/sharp/src/sharp_domain.c` line 169 at r10](https://reviewable.io/reviews/pmem/libfabric/8#-NJoLgeUBilYwAGkA0sB:-NKSTsez5R60koLpX0-_:b-896fix) ([raw file](https://github.com/pmem/libfabric/blob/f6743afdaa3c78b3303eeca1bca3b8062dc8cf33/prov/sharp/src/sharp_domain.c#L169)):*
<details><summary><i>Previously, ldorau (Lukasz Dorau) wrote…</i></summary><blockquote>

.
</blockquote></details>

Done.
___
*[`prov/sharp/src/sharp_domain.c` line 175 at r10](https://reviewable.io/reviews/pmem/libfabric/8#-NJoLhgV-AD5R-YdeJ5G:-NKSTtB6559EdpG2VLSc:b-896fix) ([raw file](https://github.com/pmem/libfabric/blob/f6743afdaa3c78b3303eeca1bca3b8062dc8cf33/prov/sharp/src/sharp_domain.c#L175)):*
<details><summary><i>Previously, ldorau (Lukasz Dorau) wrote…</i></summary><blockquote>

.
</blockquote></details>

Done.
___
*[`prov/sharp/src/sharp_ep.c` line 188 at r10](https://reviewable.io/reviews/pmem/libfabric/8#-NJoLjMDCWEPQkuX6HA6:-NKSTxgV0zpyckWFRn7g:b-896fix) ([raw file](https://github.com/pmem/libfabric/blob/f6743afdaa3c78b3303eeca1bca3b8062dc8cf33/prov/sharp/src/sharp_ep.c#L188)):*
<details><summary><i>Previously, ldorau (Lukasz Dorau) wrote…</i></summary><blockquote>

.
</blockquote></details>

Done.
___
*[`prov/sharp/src/sharp_ep.c` line 194 at r10](https://reviewable.io/reviews/pmem/libfabric/8#-NJoLkSc9I5vADiZG-9G:-NKSTyDa-U6ybr_DdW9_:b-896fix) ([raw file](https://github.com/pmem/libfabric/blob/f6743afdaa3c78b3303eeca1bca3b8062dc8cf33/prov/sharp/src/sharp_ep.c#L194)):*
<details><summary><i>Previously, ldorau (Lukasz Dorau) wrote…</i></summary><blockquote>

.
</blockquote></details>

Done.
___
*[`prov/sharp/src/sharp_eq.c` line 76 at r10](https://reviewable.io/reviews/pmem/libfabric/8#-NJoLm778G_nuiYxA3Lx:-NKSTzzu4j8IvAGeMW1k:b-896fix) ([raw file](https://github.com/pmem/libfabric/blob/f6743afdaa3c78b3303eeca1bca3b8062dc8cf33/prov/sharp/src/sharp_eq.c#L76)):*
<details><summary><i>Previously, ldorau (Lukasz Dorau) wrote…</i></summary><blockquote>

.
</blockquote></details>

Done.
___
*[`prov/sharp/src/sharp_eq.c` line 81 at r10](https://reviewable.io/reviews/pmem/libfabric/8#-NJoLnDq5Yq27FnVo0UG:-NKSU-1m8xpkwLtoYFaj:b-896fix) ([raw file](https://github.com/pmem/libfabric/blob/f6743afdaa3c78b3303eeca1bca3b8062dc8cf33/prov/sharp/src/sharp_eq.c#L81)):*
<details><summary><i>Previously, ldorau (Lukasz Dorau) wrote…</i></summary><blockquote>

.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

grom72 reviewed (commented):
> Reviewed 1 of 14 files at r11.
> *[Reviewable](https://reviewable.io/reviews/pmem/libfabric/8)* status: 11 of 37 files reviewed, 8 unresolved discussions (waiting on @ldorau)



<!-- Sent from Reviewable.io -->

ldorau reviewed (commented):
> Reviewed 3 of 14 files at r11.
> *[Reviewable](https://reviewable.io/reviews/pmem/libfabric/8)* status: 14 of 37 files reviewed, all discussions resolved (waiting on @grom72)



<!-- Sent from Reviewable.io -->

grom72 reviewed (commented):
> Reviewed 1 of 11 files at r7, 1 of 14 files at r11, 2 of 7 files at r13, 2 of 7 files at r15, 3 of 4 files at r16.
> *[Reviewable](https://reviewable.io/reviews/pmem/libfabric/8)* status: 22 of 37 files reviewed, all discussions resolved (waiting on @ldorau)



<!-- Sent from Reviewable.io -->

grom72 reviewed (commented):
> Reviewed 3 of 24 files at r1, 1 of 3 files at r2, 1 of 2 files at r3, 1 of 8 files at r6, 6 of 11 files at r7, 4 of 14 files at r11, 1 of 5 files at r12, 1 of 7 files at r15, 5 of 11 files at r17, 6 of 6 files at r18, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/pmem/libfabric/8)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @grom72)



<!-- Sent from Reviewable.io -->

grom72 reviewed (commented):
> Reviewed 1 of 6 files at r19, 5 of 5 files at r20, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/pmem/libfabric/8)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @grom72)



<!-- Sent from Reviewable.io -->
